va_explorer/va_analytics/views.py

```diff
@@ -1,8 +1,42 @@
 from django.views.generic import TemplateView
-
+from django.http import HttpResponse
+import csv
+import pandas as pd
+from django.forms.models import model_to_dict
+from verbal_autopsy.models import Location, VerbalAutopsy
 
 class DashboardView(TemplateView):
     template_name = "va_analytics/dashboard.html"
 
 
 dashboard_view = DashboardView.as_view()
+
+def download_csv(request, out_file="va_download.csv"):
+
+    # pull in va records with CODs from database 
+    valid_vas = VerbalAutopsy.objects.exclude(causes=None).prefetch_related("location").prefetch_related("causes")
+
+    # Build a location ancestors lookup and add location information at all levels to all vas
+    location_ancestors = { location.id:location.get_ancestors() for location in Location.objects.filter(location_type="facility") }
+    va_data = []
+    # extract COD and location-based fields for each va object and convert to dicts
+    for va in valid_vas:
+        # convert model object to dictionary
+        va_dict = model_to_dict(va)
+        # get location and cod
+        va_dict["location"] = va.location.name
+        va_dict["cause"] = va.causes.all()[0].cause
+        for ancestor in location_ancestors[va.location.id]:
+            va_dict[ancestor.location_type] = ancestor.name
+        va_data.append(va_dict)
+
+    va_df = pd.DataFrame.from_records(va_data)
+    # Create the HttpResponse object with the appropriate CSV header.
+    response = HttpResponse(content_type="text/csv")
+    response["Content-Disposition"] = f'attachment; filename="{out_file}"'
+    wr = csv.writer(response)
+    wr.writerow(va_df.columns.tolist())
+    wr.writerows(va_df.to_records(index=False))
```

Comment:
Why not just use the pandas to_csv function?

```
-    wr = csv.writer(response)
-    wr.writerow(va_df.columns.tolist())
-    wr.writerows(va_df.to_records(index=False))
+    va_df.to_csv(response, index=False)
```

Reply:
Done! Didn't realize you could pass the response object directly to the to_csv function.